lib/Optimizer/Transforms/StateValidation.cpp

```diff
@@ -0,0 +1,127 @@
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+#include "PassDetails.h"
+#include "cudaq/Optimizer/Builder/Intrinsics.h"
+#include "cudaq/Optimizer/Dialect/CC/CCOps.h"
+#include "cudaq/Optimizer/Dialect/Quake/QuakeOps.h"
+#include "cudaq/Optimizer/Transforms/Passes.h"
+#include "mlir/Dialect/Complex/IR/Complex.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/Passes.h"
+
+namespace cudaq::opt {
+#define GEN_PASS_DEF_STATEVALIDATION
+#include "cudaq/Optimizer/Transforms/Passes.h.inc"
+} // namespace cudaq::opt
+
+#define DEBUG_TYPE "state-validation"
+
+using namespace mlir;
+
+/// Validate that quantum code does not contain runtime calls and remove runtime
+/// function definitions.
+namespace {
+
+static bool isRuntimeStateCallName(llvm::StringRef funcName) {
+  static std::vector<const char *> names = {
+      cudaq::getCudaqState, cudaq::createCudaqStateFromDataFP32,
+      cudaq::createCudaqStateFromDataFP64, cudaq::deleteCudaqState,
+      cudaq::getNumQubitsFromCudaqState};
+  if (std::find(names.begin(), names.end(), funcName) != names.end())
+    return true;
+  return false;
+}
+
+static bool isRuntimeStateCall(Operation *callOp) {
+  if (callOp) {
+    if (auto call = dyn_cast<func::CallOp>(callOp)) {
+      if (auto calleeAttr = call.getCalleeAttr()) {
+        auto funcName = calleeAttr.getValue().str();
+        if (isRuntimeStateCallName(funcName))
+          return true;
+      }
+    }
+  }
+  return false;
+}
+
+class ValidateStateCallPattern : public OpRewritePattern<func::CallOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(func::CallOp callOp,
+                                PatternRewriter &rewriter) const override {
+    if (isRuntimeStateCall(callOp)) {
+      auto name = callOp.getCalleeAttr().getValue();
+      callOp.emitError(
+          "Synthesis did not remove func call for quantum platform: " + name);
+    }
+    return failure();
+  }
+};
+
+class ValidateStateInitPattern
+    : public OpRewritePattern<quake::InitializeStateOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(quake::InitializeStateOp initState,
+                                PatternRewriter &rewriter) const override {
+    auto stateOp = initState.getOperand(1);
+    if (isa<cudaq::cc::StateType>(stateOp.getType()))
+      initState.emitError("Synthesis did not remove `quake.init_state <veq> "
+                          "<state>` instruction");
+
+    return failure();
+  }
+};
+
+class StateValidationPass
```

Comment:
I'm skeptical this pass is really needed.

The other pass looks like it can be modified to fail if it cannot replace all instances of `quake.init_state`. And any sort of residual local definitions will already be cleaned up by MLIR passes like SymbolDCE, so we don't need to implement that.

Reply:
I don't see local definitions cleaned up, some later conversion fail because they are still there... When is SymbolDCE is invoked?

Reply:
Removed that pass